grails-app/services/org/pih/warehouse/core/DocumentTemplateService.groovy

```diff
@@ -0,0 +1,221 @@
+
+        IContext context = report.createContext();
+        context.put("order", order)
+        context.put("orderItems", orderItems);
+        context.put("orderAdjustments", orderAdjustments);
+        return context
+    }
+
+
+    def renderDocumentTemplate(Document document, Shipment shipment, OutputStream outputStream) {
+
+        InputStream inputStream = new ByteArrayInputStream(document.fileContents)
+        IXDocReport report = XDocReportRegistry.getRegistry().
+                loadReport(inputStream, TemplateEngineKind.Freemarker);
+
+        // Add properties to the context
+        IContext context = report.createContext();
+
+        def destination = [
+                name           : shipment?.destination?.name,
+                streetAddress  : shipment?.destination?.address?.address,
+                city           : shipment?.destination?.address?.city,
+                stateOrProvince: shipment?.destination?.address?.stateOrProvince,
+                postalCode     : shipment?.destination?.address?.postalCode,
+                country        : shipment?.destination?.address?.country
+        ]
+        def origin = [
+                name           : shipment?.origin?.name,
+                streetAddress  : shipment?.destination?.address?.address,
+                city           : shipment?.destination?.address?.city,
+                stateOrProvince: shipment?.destination?.address?.stateOrProvince,
+                postalCode     : shipment?.destination?.address?.postalCode,
+                country        : shipment?.destination?.address?.country
+        ]
+        def shipmentData = [
+                name                : shipment?.name,
+                shipmentNumber      : shipment?.shipmentNumber,
+                expectedShippingDate: shipment?.expectedShippingDate,
+                totalValue          : shipment?.calculateTotalValue(),
+                actualShippingDate  : shipment?.actualShippingDate,
+                status              : shipment?.currentStatus?.name
+        ]
+        context.put("origin", origin)
+        context.put("destination", destination)
+        context.put("shipment", shipmentData);
+
+        // instruct XDocReport to inspect InvoiceRow entity as well
+        // which is given as a list and iterated in a table
+        FieldsMetadata metadata = report.createFieldsMetadata();
+        //metadata.load("r", ShipmentItem.class, true);
+        metadata.addFieldAsList("r.description");
+        metadata.addFieldAsList("r.quantity");
+        metadata.addFieldAsList("r.unit");
+        metadata.addFieldAsList("r.price");
+        metadata.addFieldAsList("r.rowtotal");
+
+        def shipmentItems = shipment.shipmentItems.collect { ShipmentItem shipmentItem ->
+            [
+                    description: shipmentItem?.product?.productCode + " " + shipmentItem?.product?.name,
+                    quantity   : shipmentItem.quantity,
```

Comment:
For `description`, `unit` and `price` you have `shipmentItem?.` and for `quantity` and `rowtotal` you are using `shipmentItem.quantity`. I think it should also have a null check with `?`.

Reply:
Ignore that code for now.